docs/source/format/FlightSql.rst

```diff
@@ -0,0 +1,162 @@
+
+SQL Metadata
+------------
+
+Flight SQL provides a variety of commands to fetch catalog metadata
+about the database server.
+
+All of these commands can be used with the GetFlightInfo and GetSchema
+RPC methods. The Protobuf request message should be packed into a
+google.protobuf.Any message, then serialized and packed as the ``cmd``
+field in a CMD-type FlightDescriptor.
+
+If the command is used with GetFlightInfo, the server will return a
+FlightInfo response. The client should then use the Ticket(s) in the
+FlightInfo with the DoGet RPC method to fetch a Arrow data containing
+the results of the command. In other words, metadata is returned as
+Arrow data, just like query results themselves.
+
+The schema returned by GetSchema or DoGet for a particular command is
+fixed according to the specification.
+
+``CommandGetCatalogs``
+    List the catalogs available in the database. The definition varies
+    by vendor.
+
+``CommandGetCrossReference``
+    List the foreign key columns in a given table that reference
+    columns in a given parent table.
+
+``CommandGetDbSchemas``
+    List the schemas (note: a grouping of tables, *not* an Arrow
+    schema) available in the database. The definition varies by
+    vendor.
+
+``CommandGetExportedKeys``
+    List foreign key columns that reference the primary key columns of
+    a given table.
+
+``CommandGetImportedKeys``
+    List foreign keys of a given table.
+
+``CommandGetPrimaryKeys``
+    List the primary keys of a given table.
+
+``CommandGetSqlInfo``
+    Fetch metadata about the database server and its supported SQL
+    features.
+
+``CommandGetTables``
+    List tables in the database.
+
+``CommandGetTableTypes``
+    List table types in the database. The list of types varies by
+    vendor.
+
+Query Execution
+---------------
+
+Flight SQL also provides commands to execute SQL queries and manage
+prepared statements.
+
+Many of these commands are also used with GetFlightInfo/GetSchema and
+work identically to the metadata methods above. Some of these commands
+can be used with the DoPut RPC method, but the command should still be
+encoded in the request FlightDescriptor in the same way.
+
+Commands beginning with "Action" are instead used with the DoAction
+RPC method, in which case the command should be packed into a
+google.protobuf.Any message, then serialized and packed into the
+``body`` of a Flight Action. Also, the ``type`` should be set to the
+command name (i.e. for ``ActionClosePreparedStatementRequest``, the
+``type`` should be ``ClosePreparedStatement``).
+
+``ActionClosePreparedStatementRequest``
+    Close a previously created prepared statement.
+
+``ActionCreatePreparedStatementRequest``
+    Create a new prepared statement for a SQL query.
+
+``CommandPreparedStatementQuery``
+    Execute a previously created prepared statement and get the results.
+
+    When used with DoPut: binds parameter values to the prepared statement.
```

Comment:
So one must call first `DoPut` then `GetFlightInfo` to execute a query with parameters? Wouldn't it be nicer to be able to use `DoExchange` for that?

Reply:
I believe this was considered during the initial discussion but was rejected; @jduo do you remember why?

Reply:
Ah, it was due to output parameters potentially changing the schema mid-stream. https://github.com/apache/arrow/pull/9368#discussion_r572941765